### ISSUE TITLE
[clang][analyzer] core.NullDereference false positive for offsetof-derived field access

### ISSUE THREAD

**Issue by bozicrHT** [llvm/llvm-project#221768]
**Description:**
`q` points exactly to `s.p`, which the analyzer itself confirms by evaluating `q == &s.p` as `TRUE`. Therefore, `*q` and `s.p` must always have the same value. However, the analyzer explores a path where: `s.p == NULL` and `*q != NULL`, because of that it enters an infeasible branch and reports false positive.

**Reproducer:**
```c
#include <stddef.h>

void clang_analyzer_eval(int);
void clang_analyzer_warnIfReached(void);

struct S {
    int padding;
    int *p;
} s;

int f(void)
{
    int **q = (int **)((char *)&s + offsetof(struct S, p));

    if (!s.p && *q != NULL) {
        clang_analyzer_eval(q == &s.p); // TRUE
        clang_analyzer_eval(s.p == *q); // FALSE, should be TRUE
        clang_analyzer_warnIfReached(); // REACHABLE, should be UNREACHABLE
        return *s.p; // false positive
    }

    return 0;
}
```
Compiler explorer link can be found [here](https://compiler-explorer.com/z/6rM5rM45j).

**Comment by NagyDonat** [comment]
In many aspects, the symbolic execution of the static analyzer primarily follows the (relatively abstract and "rich") execution model of the C/C++ standard, and not the "what actually happens to the bytes" behavior of common compilers. This clashes with the `offsetof` macro, which is very traditional and well-supported, but primarily allows the programmer to do things that are undefined or disallowed by the standard.

For example, the complex [`MemRegion` class hierarchy](https://clang.llvm.org/doxygen/classclang_1_1ento_1_1MemRegion.html) distinguishes `FieldRegion` (which selects the memory area of a particular field in the parent region that represents the whole struct object) and `ElementRegion` (which primarily selects one element of an array, but is also used [abused?] for representing pointer arithmetic). Obviously, compilation would eventually translate both to byte offsets, but they are separate concepts in the standard, so the analyzer represents them separately.

I'm 80% confident that this false positive is caused by the fact that the analyzer represents `s.p` as the value stored in a `FieldRegion` subregion of the struct object, while `*q` represents the value stored in an `ElementRegion` subregion of the same struct object.

To resolve this issue (without resorting to ugly hacks) we would need to redesign the memory model of the analyzer to use a more low-level representation that explicitly uses byte offsets. This is not completely out of question (AFAIK it is on the wishlist of some experienced contributors) but is definitely a huge task.

**Comment by bozicrHT** [comment]
Thank you for the detailed explanation! Would it make sense to keep this issue open as a known false positive and long-term memory-model issue?

**Comment by NagyDonat** [comment]
Yes, personally I think this should be left open.
